### ISSUE TITLE
[java] ConstructorCallsOverridableMethodRule: false positive for method called from lambda expression

### ISSUE THREAD

**Issue by dreniers** [pmd/pmd#397]
**Rule Set:**

```
<rule ref="rulesets/java/design.xml/ConstructorCallsOverridableMethod"/>
```

**Description:**

The code below gives a false positive for the rule ConstructorCallsOverridableMethodRule for "dangerous" methods called from lambda expressions. I believe this is a false positive, because the method is not actually called during the constructor invocation. Also, replacing the lambda expressions by anonymous class does not give the violation.

**Code Sample demonstrating the issue:**

```
public class Foo {
 public Foo(String s) {
  addActionListener(() -> bar());
 }
 public void bar() {}
}
```

**Running PMD through:** *CLI*

```
C:\pmd\bin>pmd.bat -d FalsePositive.java -f xml -R rulesets/java/design.xml/ConstructorCallsOverridableMethod
```

Outputs:
```
<?xml version="1.0" encoding="UTF-8"?>
<pmd version="5.6.1" timestamp="2017-05-23T02:44:45.868">
<file name="C:\pmd\bin\FalsePositive.java">
<violation beginline="3" endline="3" begincolumn="27" endcolumn="31" rule="Const
ructorCallsOverridableMethod" ruleset="Design" class="Foo" method="Foo" external
InfoUrl="https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/design.html#Constru
ctorCallsOverridableMethod" priority="1">
Overridable method 'bar' called during object construction
</violation>
</file>
</pmd>
```

**Comment by dreniers** [comment]
I propose the fix from: https://github.com/tiobe/pmd/commit/4f826d2e080d1aa297ba7439dedd78d1bdbd46d7

I am unsure about the ramifications of setting ASTLambdaExpression.isFindBoundary() to true, but all tests still pass.

**Comment by adangel** [comment]
@dreniers Thanks for reporting!
Yes, this is indeed a false positive.

I see, there are a couple of more commits on the [pmd/5.5.x branch](https://github.com/tiobe/pmd/commits/pmd/5.3.x). Should we port them to the "official" PMD as well?

**Comment by jsotuyod** [comment]
@dreniers that's actually a very nice fix

@adangel a few of those commits are already in PMD. For instance, https://github.com/tiobe/pmd/commit/53bfc8a061f34700608d3cca02443fc1b20eeea6 was merged with #200, and https://github.com/tiobe/pmd/commit/f4e50543cf0d3d9520d0412962b7d3c159d7c16b was proposed, but rejected at #263. It would be really nice if @tiobe developers would help us merge back any other interesting changes they may have. Also, since they seem to be using 5.3.x, it may be nice to understand why they can't follow PMD, maybe we are missing something (other than not being able to run Java 7 maybe?)

**Comment by dreniers** [comment]
@jsotuyod Thanks! 

We don't have any other interesting changes in our 5.3.x branch, except for the ones already mentioned. Unfortunately, we have to stay at 5.3.x for a while because we still need to support Java 6.

Should I submit a pull request to have it merged to master?

**Comment by jsotuyod** [comment]
@dreniers please, do submit a PR with that fix. We would gladly merge it and give proper credit for the fix to you guys.
